### PR TITLE
Improve PHA source plots: support type=counts and factor above 1

### LINKED ISSUE #1825: plot_source incorrectly scaled with analysis factor >0
`plot_source` does not scaled properly Y values when `set_analysis` factor option is >0. 
The plot shows the source model plot with factors 0,1,2 and the curves are shifted in normalization, but the shape is the same.
E*f(E) should generate curves with varying shapes.

plt_source_factors

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Source plots for PHA data now follow the rate and factor analysis settings. Prior to this the rate setting was ignored and the factor setting produced incorrect values for factor=1 and factor=2). Fix #1825 

# Details

This was written separately to

- #1780 

which we had merged first. Although they are related, the source changes (needed here) are a lot simpler than needed for the model case (which needs to worry about grouping data) so the code paths are different (even though they are doing similar things). It might be nice to refactor this but I do no believe it is worth it (as the refactoring would probably be fairly extensive).

So, for factor=0, we have (assuming type=rate):

    y_0 = mdl(lo, hi) / (hi - lo)

where the model is assumed to create values of photon/cm^2/s for each bin. This means that y is in units of photon/cm^2/s/<X units> where <X units> is energy or angstrom. 

For factor = 1 and we calculate

    y_1 = y_0 * (lo + hi) / 2
    y_2 = y_1 * (lo + hi) / 2

That is, we multiply the factor_0 value by the mid-point of each X bin (and the number of times we do this depends on the factor setting).

So, for factor=0 we plot (lo, hi, y_0), factor=1 we use (lo, hi, y_1), and for factor=2 we use (lo, hi, y_2). This makes sense frm a units viewpoint, but is it what we want? For instance, for factor=1 we could just plot

    mdl(lo, hi)

instead of

    mdl(lo, hi) * (lo + hi)
    --------------------------
          2 * (hi - li)
 
Since `mdl(lo, hi)` calculates a rate, then we need to multiply by the exposure time when the user has called `set_analysis(..., type="counts")`.

This code replicates code that is in the `DataPHA` object - in partiicular the `_fix_y_units` code, but infortunately we can't just use that (to avoid code duplication) since they aren't quite doing the same thing (e.g. the source code doesn't group the data whereas the `_fix_y_units` code does, and the model being evaluated doesn't have a response whereas it does for the DataPHA case. So I think we have to live with this "almost repeated code".

### CODE REVIEW

**Comment by DougBurke** [comment]
@anetasie - I think this shoulf "fix" your issue #1825, but it needs some careful testing (e.g. to look at the x/y values you get and whether they make sense)

**Comment by anetasie** [comment]
@DougBurke Thanks. This fixes the problems for set_analysis(1,'energy','rate',factor=???) and it looks good.
I'm wondering about 'counts' option, which generates the plots with really high values and as you said the Y value needs to be corrected by the exposure.

**Comment by anetasie** [comment]
I'm also wondering about the translation to wavelength and rate. Need to get some grating spectra with wavelength scales for testing.

**Comment by DougBurke** [comment]
So there's

    set_analysis(idval, "energy" | "wave", "rate", factor=0, 1, 2, ...)

which is the main one. I also added support for

    set_analysis(idval, "energy" | "wave", "counts", factor=0, 1, 2, ...)

but only because we can - it isn't particularly useful that I can see (it's not really counts, but I think that is more a problem with this argument, it should have just said "do I want a rate or not"). I would leave it in, but if we feel it should be removed it's not hard to do so.

**Comment by anetasie** [comment]
@DougBurke which powerlaw model are you using?

**Comment by anetasie** [comment]
I guess my question is related to the definition of the photon index gamma. For factor 2 in powlaw1d I expect the function rising up. This is what I tested and it looked ok

**Comment by anetasie** [comment]
OK the plots look good. I just got confused with gamma vs. factors.

**Review by anetasie — APPROVED** [review]

**Comment by DougBurke** [comment]
@hamogu - I've asked for your review just to check the code isn't too crazy, and because you have been thinking about the plotting code recently too.

**Comment by DougBurke** [comment]
I've rebased to deal with a minor conflict from merging #1780 which gave me the chance to combine two commits and fix at least one typo in a comment.

**Comment by DougBurke** [comment]
Here's the script I used for the plots above:

```
import matplotlib.pyplot as plt

from sherpa.astro.ui import *

load_pha("sherpa-test-data/sherpatest/3c273.pi")
notice(0.5, 6)

set_source(xsphabs.gal * powlaw1d.pl)

pl.ampl = 2e-5

def make_plot(units, factor):

    set_analysis(1, units, factor=factor)

    pl.gamma = 2
    plot_source_component(pl, xlog=True, ylog=True)
    ax = plt.gca()
    ax.lines[0].set_label("gamma = 2")

    pl.gamma = 1
    plot_source_component(pl, overplot=True)
    ax.lines[-2].set_label("gamma = 1")

    pl.gamma = 0
    plot_source_component(pl, overplot=True)
    ax.lines[-2].set_label("gamma = 0")

    pl.gamma = -1
    plot_source_component(pl, overplot=True)
    ax.lines[-2].set_label("gamma = -1")

    pl.gamma = -2
    plot_source_component(pl, overplot=True)
    ax.lines[-2].set_label("gamma = -2")

    plt.legend()
    plt.title(f"units={units}  factor={factor}")

    out = f"plot-{units}-{factor}.png"
    plt.savefig(out)
    print(f"Created: {out}")


for unit in ["energy", "wave"]:
    for factor in [0, 1, 2]:
        make_plot(unit, factor)
```

File: sherpa/astro/tests/test_astro_plot.py [hunk 1]
```diff
@@ -48,20 +48,28 @@ def check_sourceplot_energy(sp, factor=0):
     assert sp.xlabel == 'Energy (keV)'
 
     # the following depends on the backend
-    # assert sp.ylabel == 'f(E)  Photons/sec/cm$^2$/keV'
     if factor == 0:
-        assert sp.ylabel.startswith('f(E)  Photons/sec/cm')
-        assert sp.ylabel.endswith('/keV ')
+        if rate:
+            assert sp.ylabel.startswith('f(E)  Photons/sec/cm')
+        else:
```
Comment (hamogu) [inline sherpa/astro/tests/test_astro_plot.py:54]: Line 42 sets "rate=True". Was this intended to be parameterized for rate=True/False?

File: sherpa/astro/tests/test_astro_plot.py [hunk 2]
```diff
@@ -95,18 +108,27 @@ def check_sourceplot_wavelength(sp, factor=0):
     assert sp.xlabel == 'Wavelength (Angstrom)'
 
     if factor == 0:
-        assert sp.ylabel.startswith('f(lambda)  Photons/sec/cm')
-        assert sp.ylabel.endswith('/Angstrom ')
+        if rate:
```
Comment (hamogu) [inline sherpa/astro/tests/test_astro_plot.py:111]: Same. rate is always True in this test.

**Review by hamogu — CHANGES_REQUESTED** [review]

File: sherpa/astro/tests/test_astro_plot.py [hunk 3]
```diff
@@ -195,7 +254,7 @@ def test_sourceplot_counts(caplog, make_basic_datapha):
         sp.prepare(data, src)
 
     assert len(caplog.records) == 0
-    check_sourceplot_energy(sp)
+    check_sourceplot_energy(sp, rate=False)
```
Comment (DougBurke) [inline sherpa/astro/tests/test_astro_plot.py:257]: @hamogu - here is a case when `check_sourceplot_energy` is called with `rate=False`.

File: sherpa/astro/tests/test_astro_plot.py [hunk 4]
```diff
@@ -330,7 +417,67 @@ def test_sourceplot_wavelength_counts(caplog, make_basic_datapha):
         sp.prepare(data, src)
 
     assert len(caplog.records) == 0
-    check_sourceplot_wavelength(sp)
+    check_sourceplot_wavelength(sp, rate=False)
```
Comment (DougBurke) [inline sherpa/astro/tests/test_astro_plot.py:420]: @hamogu and ditto for `check_sourceplot_wavelength`.

File: sherpa/astro/tests/test_astro_plot.py (hunk 2 above)
Comment (hamogu) [reply sherpa/astro/tests/test_astro_plot.py:111]: Then why does this say"if rate"?

**Comment by hamogu** [comment]
I'm sorry,  I missed that. On Aug 22, 2023 10:37 AM, Doug Burke ***@***.***> wrote:
@DougBurke commented on this pull request.



In sherpa/astro/tests/test_astro_plot.py:
> @@ -95,18 +108,27 @@ def check_sourceplot_wavelength(sp, factor=0):
     assert sp.xlabel == 'Wavelength (Angstrom)'
 
     if factor == 0:
-        assert sp.ylabel.startswith('f(lambda)  Photons/sec/cm')
-        assert sp.ylabel.endswith('/Angstrom ')
+        if rate:

I don't understand.
This is not a test itself - it's name doesn't begin with test_ - but is used in tests. It defaults to rate=True but is also called with rate=False - e.g. it is called below at line 420 / see https://github.com/sherpa/sherpa/pull/1828/files#r1300537291
This is also true of the other case you are asking about.

**Review by hamogu — APPROVED** [review]

**Comment by hamogu** [comment]
I'm sorry I missed the true/ false thing.  I was reviewing changes on my phone.

**Comment by DougBurke** [comment]
... the perils of PR reviewing on a beach ....

**Pull request merged**